Dust protection

alexsporn opened the pull request:
This PR add dust protection as defined in the RFC and implemented in iota.go.

*Warning:* this changes the snapshot format to add the output type for each output.

**New:**

- Dust-validation during white-flag confirmation
- New conflict reason if transaction is ignored due to invalid dust allowance
- Dust allowance tracking is done in the UTXOManager with support for applying and rolling back whole confirmations diffs. This means that Dust is calculated properly when a node is started from a snapshot.

Note: the dust allowance tracking is done twice in this PR.
1. Inside the white-flag confirmation to be able to validate the semantics
2. After white-flag when applying the end-result of the white-flag confirmation.

Why?
- Temporary dust that exists between milestones is never booked into the ledger database.
- By having the UTXOManager calculate the dust internally we automatically gain the support for snapshots.

alexsporn committed:
- Added new Dust database to store dust stats for each address

- Updated to latest iota.go and 3rdparty dependencies

alexsporn committed:
- Add output type to snapshot file (this is breaking)

- Use Output interface to get the Type, Deposit and Target instead of hardcoding the SigLockedSingleOutput

alexsporn committed:
- Add dust protection steps to white-flag and confirmation

alexsporn committed:
- Moved the Dust storage to the UTXOManager

- Keep the collected dust stats during white-flag internal and do not apply them on confirmation

alexsporn committed:
- Added missing spent dust removal from white-flag computation

alexsporn committed:
- Added dust calculation for milestone diffs

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+
+func (u *Manager) ReadDustForAddress(address iotago.Address) (dustAllowanceBalance uint64, dustOutputCount int64, err error) {
+
+	return u.readDustForAddress([]byte(address.String()))
```
this doesn't seem right to me, you are using the hex encode string instead of the address bytes here.

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+func (u *Manager) storeDustForAddress(addressBytes []byte, dustAllowanceBalance uint64, dustOutputCount int64, mutations kvstore.BatchedMutations) error {
+
+	if dustOutputCount == 0 && dustAllowanceBalance != 0 {
+		// Balance cannot be zero if there are no outputs
```
"be non-zero"

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+	if dustAllowanceBalance == 0 {
+		// Remove from database
+		return mutations.Delete(addressBytes)
+	} else {
```
Why "else" here?
The former "if" already returns.

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+		return fmt.Errorf("%w: %s dustAllowanceBalance %d, dustOutputCount %d", ErrInvalidDustForAddress, hex.EncodeToString(addressBytes), dustAllowanceBalance, dustOutputCount)
+	}
+
+	if dustAllowanceBalance == 0 {
```
If you delete if the count is 0, you could split up the former if:

```go

	if dustOutputCount == 0 {
		if dustAllowanceBalance != 0 {
			// Balance cannot be non-zero if there are no outputs
			return fmt.Errorf("%w: %s dustAllowanceBalance %d, dustOutputCount %d", ErrInvalidDustForAddress, hex.EncodeToString(addressBytes), dustAllowanceBalance, dustOutputCount)
		}

		// Remove from database
		return mutations.Delete(addressBytes)
	}
```

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+
+func (u *Manager) applyDustDiffForAddress(addressString string, dustAllowanceBalanceDiff int64, dustOutputCountDiff int64, mutations kvstore.BatchedMutations) error {
+
+	addressBytes := []byte(addressString)
```
Better call it `addressMapKey`

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+		// Add new dust
+		switch output.outputType {
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			address := output.Address().String()
```
you should convert the address bytes to string, but the String function uses hex encoding

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			address := output.Address().String()
+			diff, found := dustDiff[address]
+			if found {
```
Can you move that "if" in the line above?

```go

			if diff, found := dustDiff[address]; found {
				diff.DustAllowanceBalanceDiff += int64(output.Amount())
			} else {
				dustDiff[address] = NewDustDiff(int64(output.Amount()), 0)
			}
```

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+		case iotago.OutputSigLockedSingleOutput:
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				address := output.Address().String()
+				diff, found := dustDiff[address]
```
Can you move that "if" in the line above?

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+			}
+		case iotago.OutputSigLockedSingleOutput:
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				address := output.Address().String()
```
you should convert the address bytes to string, but the String function uses hex encoding

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+		output := spent.Output()
+		switch output.outputType {
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			address := output.Address().String()
```
you should convert the address bytes to string, but the String function uses hex encoding

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			address := output.Address().String()
+			diff, found := dustDiff[address]
+			if found {
```
Can you move that "if" in the line above?

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				address := output.Address().String()
+				diff, found := dustDiff[address]
+				if found {
```
Can you move that "if" in the line above?

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			address := output.Address().String()
+			diff, found := dustDiff[address]
+			if found {
```
Can you move that "if" in the line above?

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				address := output.Address().String()
+				diff, found := dustDiff[address]
+				if found {
```
Can you move that "if" in the line above?

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			address := output.Address().String()
+			diff, found := dustDiff[address]
+			if found {
```
Can you move that "if" in the line above?

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				address := output.Address().String()
+				diff, found := dustDiff[address]
+				if found {
```
Can you move that "if" in the line above?

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+			}
+		case iotago.OutputSigLockedSingleOutput:
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				address := output.Address().String()
```
you should convert the address bytes to string, but the String function uses hex encoding

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+		// Remove unspent dust
+		switch output.outputType {
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			address := output.Address().String()
```
you should convert the address bytes to string, but the String function uses hex encoding

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+			}
+		case iotago.OutputSigLockedSingleOutput:
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				address := output.Address().String()
```
you should convert the address bytes to string, but the String function uses hex encoding

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+		output := spent.Output()
+		switch output.outputType {
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			address := output.Address().String()
```
you should convert the address bytes to string, but the String function uses hex encoding

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+			}
+		case iotago.OutputSigLockedSingleOutput:
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				address := output.Address().String()
```
you should convert the address bytes to string, but the String function uses hex encoding

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+
+	switch unspentOutput.outputType {
+	case iotago.OutputSigLockedDustAllowanceOutput:
+		if err := u.applyDustDiffForAddress(unspentOutput.Address().String(), int64(unspentOutput.Amount()), 0, dustMutations); err != nil {
```
you should convert the address bytes to string, but the String function uses hex encoding

muXxer commented on pkg/model/utxo/dust_db.go:
```diff
@@ -0,0 +1,223 @@
+		}
+	case iotago.OutputSigLockedSingleOutput:
+		if unspentOutput.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+			if err := u.applyDustDiffForAddress(unspentOutput.Address().String(), 0, 1, dustMutations); err != nil {
```
you should convert the address bytes to string, but the String function uses hex encoding

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -178,8 +181,29 @@ func ComputeWhiteFlagMutations(s *storage.Storage, msIndex milestone.Index, cach
+			}
+
+			dustDiff, found := wfConf.dustDiff[address.String()]
+			if found {
```
Can you move that "if" in the line above?

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -231,11 +257,55 @@ func ComputeWhiteFlagMutations(s *storage.Storage, msIndex milestone.Index, cach
+			case iotago.OutputSigLockedDustAllowanceOutput:
+				address := input.Address().String()
+				dustDiff, found := wfConf.dustDiff[address]
+				if found {
```
Can you move that "if" in the line above?

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -231,11 +257,55 @@ func ComputeWhiteFlagMutations(s *storage.Storage, msIndex milestone.Index, cach
+				if input.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+					address := input.Address().String()
+					dustDiff, found := wfConf.dustDiff[address]
+					if found {
```
Can you move that "if" in the line above?

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -231,11 +257,55 @@ func ComputeWhiteFlagMutations(s *storage.Storage, msIndex milestone.Index, cach
+			case iotago.OutputSigLockedDustAllowanceOutput:
+				address := output.Address().String()
+				dustDiff, found := wfConf.dustDiff[address]
+				if found {
```
Can you move that "if" in the line above?

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -231,11 +257,55 @@ func ComputeWhiteFlagMutations(s *storage.Storage, msIndex milestone.Index, cach
+				if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+					address := output.Address().String()
+					dustDiff, found := wfConf.dustDiff[address]
+					if found {
```
Can you move that "if" in the line above?

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -231,11 +257,55 @@ func ComputeWhiteFlagMutations(s *storage.Storage, msIndex milestone.Index, cach
+				}
+			case iotago.OutputSigLockedSingleOutput:
+				if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+					address := output.Address().String()
```
you should convert the address bytes to string, but the String function uses hex encoding

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -231,11 +257,55 @@ func ComputeWhiteFlagMutations(s *storage.Storage, msIndex milestone.Index, cach
+			// Add new dust to diff
+			switch output.OutputType() {
+			case iotago.OutputSigLockedDustAllowanceOutput:
+				address := output.Address().String()
```
you should convert the address bytes to string, but the String function uses hex encoding

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -231,11 +257,55 @@ func ComputeWhiteFlagMutations(s *storage.Storage, msIndex milestone.Index, cach
+				}
+			case iotago.OutputSigLockedSingleOutput:
+				if input.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+					address := input.Address().String()
```
you should convert the address bytes to string, but the String function uses hex encoding

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -231,11 +257,55 @@ func ComputeWhiteFlagMutations(s *storage.Storage, msIndex milestone.Index, cach
+			// Subtract spent dust from diff
+			switch input.OutputType() {
+			case iotago.OutputSigLockedDustAllowanceOutput:
+				address := input.Address().String()
```
you should convert the address bytes to string, but the String function uses hex encoding

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -178,8 +181,29 @@ func ComputeWhiteFlagMutations(s *storage.Storage, msIndex milestone.Index, cach
+				return 0, 0, err
+			}
+
+			dustDiff, found := wfConf.dustDiff[address.String()]
```
you should convert the address bytes to string, but the String function uses hex encoding

alexsporn committed:
- Requested changes

alexsporn committed:
- Fix value test by sending enough coins for a valid deposit

alexsporn committed:
- Created new DustAllowanceDiff type that takes outputs and spents to calculate the diff

- Now using the serialized iota addresses (they include a type, so not just the raw bytes)

alexsporn committed:
- Remove else since we return inside the if

alexsporn committed:
- Simplify if

alexsporn committed:
- Ignore Goland files

alexsporn committed:
- Updated iota.go to latest

alexsporn committed:
- Mark the message containing a milestone in the metadata and add `milestoneIndex` to the metadata API and MQTT payloads if the message contains the milestone

alexsporn committed:
- Do not lock the UTXO ledger when accessing the API since the information can be old anyways when it reaches the client

muXxer commented on pkg/whiteflag/confirmation.go:
```diff
@@ -168,6 +167,23 @@ func ConfirmMilestone(s *storage.Storage, serverMetrics *metrics.ServerMetrics,
+		meta.GetMetadata().SetIsNoTransaction(true)
+		if !meta.GetMetadata().IsReferenced() {
+			meta.GetMetadata().SetReferenced(true, milestoneIndex)
+			meta.GetMetadata().SetMilestone(true)
```
This way we would only mark the first message with the milestone payload we receive as milestone.
This could differ from node to node. But I don't know how to avoid it :/

alexsporn replied:
Yeah, but it is the one we actually used to process the milestone for this node, so better than nothing 🤷

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -52,6 +52,8 @@ type WhiteFlagMutations struct {
 	NewOutputs map[string]*utxo.Output
 	// Contains the Spent Outputs for the given confirmation.
 	NewSpents map[string]*utxo.Spent
+	// Contains the Dust allowance diff for the given addresses.
```
is "for the given address" correct in this context?

muXxer commented on pkg/model/utxo/dust.go:
```diff
@@ -0,0 +1,180 @@
+		switch output.outputType {
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			// Add new dust allowance
+			addressKey := output.Address().String()
```
this seems wrong

muXxer commented on pkg/model/utxo/dust.go:
```diff
@@ -0,0 +1,180 @@
+		case iotago.OutputSigLockedSingleOutput:
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				// Add new dust
+				addressKey := output.Address().String()
```
this seems wrong

alexsporn committed:
- Fixed review comments

alexsporn committed:
- Add the output type to the binary snapshot file

alexsporn committed:
- Regenerated snapshot files for the integration tests

alexsporn committed:
- Moved storage.LoadInitialValuesFromDatabase to when the storage is created instead of requiring it to be called from the outside

alexsporn committed:
- Fixed variable shadowing in NewOutput()

alexsporn committed:
- Added CachedMilestones type

- Added a new UTXOInput() method to the utxo.Output

alexsporn committed:
- Fixed and modernized the Testsuite and the whiteflag tests

alexsporn committed:
- Added AssertMessageConflictReason to testsuite

alexsporn committed:
- Added first Dust test

alexsporn committed:
- Add support for creating dust allowance in the testsuite

alexsporn committed:
- Fix dust allowance reading failing for addresses that do not have any dust

alexsporn committed:
- Add dust allowance to dust tests

alexsporn committed:
- Removed unused field from confirmation stats

alexsporn committed:
- Expanded dust tests and fixed InputToOutputMapping() not setting the correct output type

alexsporn committed:
- Sweep all dust+allowance in test. (This requires an iota.go fix that still needs to be merged)

alexsporn committed:
- Adds dust test that converts a single output into a dust allowance on the same address.

- Fix in the Dust ledger

alexsporn committed:
- Test hitting the dust limit and then increasing it

alexsporn commented:
This needs https://github.com/iotaledger/iota.go/pull/238 to be merged first and go.mod updated

alexsporn committed:
- Updated iota.go to latest version

alexsporn committed:
- Refactored testsuite with a new builder pattern that reduces code duplication and improves readability

alexsporn committed:
- Move NewMessageBuilder to the TestEnvironment

alexsporn committed:
- Increase amount in snapshot integration tests, so that they do not fail due to being dust. Also added the output type to the tool generating the snapshots, although it would have defaulted to the correct value (0).

alexsporn committed:
- Removed separate dust storage from UTXOManager and instead use the normal utxoStorage with a dust prefix. This way we write dust in the same batched mutation as the UTXO itself.

alexsporn committed:
- Removed unnecessary value not nil check

alexsporn committed:
- Send proper raw output type in the rest API

alexsporn committed:
- Send proper raw output type over MQTT

alexsporn committed:
- Updated iota.go with output parsing fix
